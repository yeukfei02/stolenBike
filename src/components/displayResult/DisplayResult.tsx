import React from 'react';
import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';
import Grid from '@material-ui/core/Grid';
import Paper from '@material-ui/core/Paper';
import Typography from '@material-ui/core/Typography';
import ButtonBase from '@material-ui/core/ButtonBase';
import { useHistory } from 'react-router-dom';
import moment from 'moment';

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      flexGrow: 1,
    },
    paper: {
      padding: theme.spacing(2),
      margin: '2em auto',
      maxWidth: 1000,
    },
    image: {
      width: 128,
      height: 128,
      border: '0.1em black solid',
    },
    img: {
      margin: 'auto',
      display: 'block',
      maxWidth: '100%',
      maxHeight: '100%',
    },
  }),
);

function DisplayResult(props: any) {
  const classes = useStyles();
  const history = useHistory();

  const handlePaperViewTitleClick = (id: string) => {
    history.push(`/case/${id}`);
  };

  const renderDisplayResult = () => {
    let paperViewResultList = null;

    if (props.resultList) {
      paperViewResultList = props.resultList.map((item: any, i: number) => {
        const occurredDate = moment(item.occurred_at).format("ddd MMM DD YYYY");

        return (
          <Paper key={i} className={classes.paper}>
            <Grid container spacing={2}>
              <Grid item>
                <ButtonBase className={classes.image}>
                  <img className={classes.img} alt="" src={item.media.image_url_thumb} />
                </ButtonBase>
              </Grid>
              <Grid item xs={12} sm container>
                <Grid item xs container direction="column" spacing={2}>
                  <Grid item xs>
                    <Typography
                      gutterBottom
                      variant="subtitle1"
                      style={{ color: 'blue', textDecoration: 'underline', cursor: 'pointer' }}
                      onClick={() => handlePaperViewTitleClick(item.id)}
                    >
                      {item.title}
                    </Typography>
                    <Typography variant="body2" gutterBottom>
                      {item.description}
                    </Typography>
                    <Typography variant="body2" color="textSecondary">
                      {occurredDate} - {item.address}
                    </Typography>
                  </Grid>
                </Grid>
              </Grid>
            </Grid>
          </Paper>
        );
      });
    }

    return paperViewResultList;
  };

  return <div className={classes.root}>{renderDisplayResult()}</div>;
}

export default DisplayResult;
