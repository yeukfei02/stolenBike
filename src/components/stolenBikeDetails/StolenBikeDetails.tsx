import React, { useState, useEffect } from 'react';
import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';
import Paper from '@material-ui/core/Paper';
import Button from '@material-ui/core/Button';
import { withRouter } from 'react-router';
import { useHistory } from 'react-router-dom';
import _ from 'lodash';
import axios from 'axios';
import moment from 'moment';

import CustomMap from '../customMap/CustomMap';

const ROOT_URL = 'https://bikewise.org/api/v2';

async function getLocationsMarkers() {
  const response = await axios.get(`${ROOT_URL}/locations/markers`);
  return response;
}

async function getIncidentById(id: string) {
  const response = await axios.get(`${ROOT_URL}/incidents/${id}`);
  return response;
}

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      padding: theme.spacing(3, 2),
    },
  }),
);

function StolenBikeDetails(props: any) {
  const classes = useStyles();
  const history = useHistory();

  const [latitude, setLatitude] = useState<number>(0);
  const [longitude, setLongitude] = useState<number>(0);

  const [stolenBikeDetails, setStolenBikeDetails] = useState<any>(null);

  const id = props.match.params.id;

  useEffect(() => {
    if (id) {
      const locationsMarkersResponse = getLocationsMarkers();
      locationsMarkersResponse
        .then((response) => {
          if (response && response.data) {
            response.data.features.forEach((item: any, i: number) => {
              if (_.isEqual(item.properties.id.toString(), id) && item.geometry && item.geometry.coordinates) {
                const latitude = item.geometry.coordinates[1];
                const longitude = item.geometry.coordinates[0];
                setLatitude(latitude);
                setLongitude(longitude);
              }
            });
          }
        })
        .catch((e) => {
          console.log('error = ', e.message);
        });

      const result = getIncidentById(id);
      result
        .then((response) => {
          if (response && response.data) setStolenBikeDetails(response.data.incident);
        })
        .catch((e) => {
          console.log('error = ', e.message);
        });
    }
  }, [id]);

  const handleBack = () => {
    history.push(`/`);
  };

  const renderStolenBikeDetailsDiv = () => {
    let stolenBikeDetailsDiv = null;

    if (stolenBikeDetails) {
      const occurredDate = moment.unix(stolenBikeDetails.occurred_at).format('MMM Do, ka');

      stolenBikeDetailsDiv = (
        <div className="my-3">
          <div className="h5">
            <b>{stolenBikeDetails.title}</b>
          </div>
          <div className="h6">
            <span>
              <b>Stolen</b>
            </span>{' '}
            {occurredDate}
            <span>
              {' '}
              <b>at</b> {stolenBikeDetails.address}
            </span>
          </div>
          <CustomMap
            latitude={latitude}
            longitude={longitude}
            name={stolenBikeDetails.title}
            address={stolenBikeDetails.address}
          />
          <div className="mt-3">
            <div className="h5">
              <b>DESCRIPTION OF INCIDENT</b>
            </div>
            <div className="h6">{stolenBikeDetails.description}</div>
          </div>
          <Button className="mt-2" variant="contained" color="primary" size="large" onClick={handleBack}>
            Back
          </Button>
        </div>
      );
    }

    return stolenBikeDetailsDiv;
  };

  return (
    <div className={classes.root}>
      <div className="mt-5 mb-3 d-flex justify-content-center">
        <Paper className={`${classes.root} w-75 mx-4`}>
          <div className="h2">
            <b>Police Department</b>
          </div>
          <div className="my-3 h3">Stolen Bike</div>
          {renderStolenBikeDetailsDiv()}
        </Paper>
      </div>
    </div>
  );
}

export default withRouter(StolenBikeDetails);
