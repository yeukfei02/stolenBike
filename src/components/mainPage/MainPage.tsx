import React, { useState, useEffect } from 'react';
import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';
import Paper from '@material-ui/core/Paper';
import Grid from '@material-ui/core/Grid';
import TextField from '@material-ui/core/TextField';
import DateFnsUtils from '@date-io/date-fns';
import 'date-fns';
import {
  MuiPickersUtilsProvider,
  KeyboardDatePicker,
} from '@material-ui/pickers';
import Button from '@material-ui/core/Button';
import axios from 'axios';
import moment from 'moment';
import _ from 'lodash';

import DisplayResult from '../displayResult/DisplayResult';

const ROOT_URL = "https://bikewise.org/api/v2";

async function getIncidents(queryData: any) {
  const response = await axios.get(`${ROOT_URL}/incidents`,
    {
      params: queryData
    }
  );
  return response;
}

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      padding: theme.spacing(3, 2),
    },
  }),
);

function MainPage() {
  const classes = useStyles();

  const [caseDescription, setCaseDescription] = useState<string | ''>('');
  const [fromDate, setFromDate] = useState<Date | null>(null);
  const [toDate, setToDate] = useState<Date | null>(null);

  const [resultList, setResultList] = useState<any[]>([]);
  const [status, setStatus] = useState<string>('');

  useEffect(() => {
    if (resultList)
      setStatus('');
  }, [resultList]);

  const handleSearchCaseDescriptionChange = (e: any) => {
    setCaseDescription(e.target.value);
  }

  const handleFromDateChange = (date: Date | null) => {
    setFromDate(date);
  };

  const handleToDateChange = (date: Date | null) => {
    setToDate(date);
  };

  const handleFindCasesClick = async () => {
    setStatus('loading');

    let formDateNum = 0;
    let toDateNum = 0;

    if (fromDate) {
      const formattedFormDate = moment(fromDate).format("X");
      formDateNum = parseInt(formattedFormDate, 10);
    }
    if (toDate) {
      const formattedToDate = moment(toDate).format("X");
      toDateNum = parseInt(formattedToDate, 10);
    }

    let queryData = {
      page: 1,
      per_page: 10,
    };
    if (caseDescription) {
      let obj = {
        query: ''
      };
      obj.query = caseDescription;
      queryData = Object.assign(queryData, obj);
    }
    if (formDateNum) {
      let obj = {
        occurred_after: 0,
      };
      obj.occurred_after = formDateNum;
      queryData = Object.assign(queryData, obj);
    }
    if (toDateNum) {
      let obj = {
        occurred_before: 0,
      };
      obj.occurred_before = toDateNum;
      queryData = Object.assign(queryData, obj);
    }

    try {
      const result = await getIncidents(queryData);
      setResultList(result.data.incidents);
    } catch (e) {
      console.log("error = ", e.message);
      setStatus('error');
    }
  }

  const renderResultDiv = () => {
    let resultDiv = null;

    if (!_.isEmpty(status)) {
      if (_.isEqual(status, 'loading')) {
        resultDiv = (
          <div className={classes.root}>
            <div className="d-flex justify-content-center">
              <Paper className={`${classes.root} w-75 mx-4`}>
                <div className="h5">Loading...</div>
              </Paper>
            </div>
          </div>
        );
      } else if (_.isEqual(status, 'error')) {
        resultDiv = (
          <div className={classes.root}>
            <div className="d-flex justify-content-center">
              <Paper className={`${classes.root} w-75 mx-4`}>
                <div className="h5" style={{ color: 'red' }}>Oops, something went wrong</div>
              </Paper>
            </div>
          </div>
        );
      }
    } else {
      if (!_.isEmpty(resultList)) {
        resultDiv = (
          <DisplayResult resultList={resultList} />
        );
      } else {
        resultDiv = (
          <div className={classes.root}>
            <div className="d-flex justify-content-center">
              <Paper className={`${classes.root} w-75 mx-4`}>
                <div className="h5">No reuslts</div>
              </Paper>
            </div>
          </div>
        );
      }
    }

    return resultDiv;
  }

  return (
    <div className={classes.root}>
      <div className="mt-5 mb-3 d-flex justify-content-center">
        <Paper className={`${classes.root} w-75 mx-4`}>
          <div className="h2"><b>Police Department</b></div>
          <div className="h4">Stolen Bike</div>
          <Grid container spacing={3}>
            <Grid item xs={6} sm={3}>
              <TextField
                id="outlined-full-width"
                label="Search case descriptions"
                placeholder="Search case descriptions"
                fullWidth
                margin="normal"
                variant="outlined"
                InputLabelProps={{
                  shrink: true,
                }}
                value={caseDescription}
                onChange={handleSearchCaseDescriptionChange}
              />
            </Grid>
            <Grid item xs={6} sm={3}>
              <MuiPickersUtilsProvider utils={DateFnsUtils}>
                <KeyboardDatePicker
                  margin="normal"
                  id="from"
                  label="from"
                  format="yyyy/MM/dd"
                  value={fromDate}
                  onChange={handleFromDateChange}
                  KeyboardButtonProps={{
                    'aria-label': 'from date',
                  }}
                  autoOk={true}
                />
              </MuiPickersUtilsProvider>
            </Grid>
            <Grid item xs={6} sm={3}>
              <MuiPickersUtilsProvider utils={DateFnsUtils}>
                <KeyboardDatePicker
                  margin="normal"
                  id="to"
                  label="to"
                  format="yyyy/MM/dd"
                  value={toDate}
                  onChange={handleToDateChange}
                  KeyboardButtonProps={{
                    'aria-label': 'to date',
                  }}
                  autoOk={true}
                />
              </MuiPickersUtilsProvider>
            </Grid>
            <Grid item xs={6} sm={3}>
              <Button className="mt-2 w-100 h-75" variant="contained" color="primary" size="large" onClick={handleFindCasesClick}>
                Find cases
              </Button>
            </Grid>
          </Grid>
        </Paper>
      </div>
      {renderResultDiv()}
    </div>
  );
}

export default MainPage;
