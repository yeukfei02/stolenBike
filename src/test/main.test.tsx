import React from 'react';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, mount } from 'enzyme';

import MainPage from '../components/mainPage/MainPage';
import DisplayResult from '../components/displayResult/DisplayResult';
import StolenBikeDetails from '../components/stolenBikeDetails/StolenBikeDetails';

Enzyme.configure({ adapter: new Adapter() });

jest.mock('react-router-dom', () => ({
  useHistory: () => ({
    push: jest.fn(),
  }),
}));

describe('main.test', () => {
  describe('render test', () => {
    it('MainPage', () => {
      shallow(<MainPage />);
    });

    it('DisplayResult', () => {
      shallow(<DisplayResult />);
    });

    it('StolenBikeDetails', () => {
      shallow(<StolenBikeDetails />);
    });
  });
});
