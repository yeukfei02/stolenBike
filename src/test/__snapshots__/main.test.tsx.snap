// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`main.test render test CustomMap 1`] = `ShallowWrapper {}`;

exports[`main.test render test DisplayResult 1`] = `ShallowWrapper {}`;

exports[`main.test render test MainPage 1`] = `ShallowWrapper {}`;

exports[`main.test render test StolenBikeDetails 1`] = `ShallowWrapper {}`;
